compiler_base/tools/src/test/task/mod.rs

```diff
@@ -0,0 +1,342 @@
+//! This file provides everything to define a [`Task`] that an [`crate::executor::Executor`] can execute.
+use std::{fmt, sync::mpsc::Sender, thread};
+pub mod event;
+
+/// [`Task`] is the unit that [`crate::executor::Executor`] can execute concurrently.
+///
+/// # Example
+///
+/// ```rust
+/// use compiler_base_tools::test::task::Task;
+/// use compiler_base_tools::test::task::FinishedTask;
+/// use std::sync::mpsc::channel;
+/// use std::sync::mpsc::Sender;
+/// use compiler_base_tools::test::task::TaskName;
+/// use compiler_base_tools::test::task::TaskId;
+/// use compiler_base_tools::test::task::TaskInfo;
+/// use compiler_base_tools::test::task::TaskStatus;
+///
+/// // 1. Define a custom task [`MyTask`].
+/// struct MyTask {
+///     id: usize,
+///     name: String,
+/// }
+///
+/// // 2. Implement trait [`Task`] for [`MyTask`].
+/// impl Task for MyTask {
+///     fn run(&self, ch: Sender<FinishedTask>) {
+///         // [`FinishedTask`] is constructed here passed to other threads via [`ch`].
+///         let res = FinishedTask::new(self.info(), vec![], vec![], TaskStatus::Finished);
+///         ch.send(res).unwrap();
+///     }
+///
+///     fn info(&self) -> TaskInfo {
+///         TaskInfo::new(self.id.into(), self.name.to_string().into())
+///     }
+/// }
+///
+/// impl MyTask {
+///     pub fn new(id: usize, name: String) -> Self {
+///         Self { id, name }
+///    }
+/// }
+///
+/// // 3. Create [`channel`] to pass [`FinishedTask`].
+/// let (tx, rx) = channel::<FinishedTask>();
+/// let my_task = MyTask::new(0, "MyTask 0".to_string());
+/// my_task.run(tx);
+///
+/// // 4. [`FinishedTask`] created in [`Task`] will be got from channel.
+/// match rx.recv() {
+///     Ok(res) => {
+///         assert_eq!(res.tinfo().tid(), 0.into());
+///         assert_eq!(res.tinfo().tname(), "MyTask 0".to_string().into());
+///         assert_eq!(res.status(), TaskStatus::Finished);
+///     },
+///     Err(_) => panic!("unreachable code")
+/// }
+/// ```
+pub trait Task {
+    /// [`run`] will be executed of the [`Task`],
+    /// and the result of the execution is communicated with other threads through the [`ch`] which is a [`Sender<FinishedTask>`],
+    /// so [`run`] method does not need to return a value.
+    fn run(&self, ch: Sender<FinishedTask>);
+
+    /// Return the [`TaskInfo`]
+    fn info(&self) -> TaskInfo;
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct TaskInfo {
```

Comment:
What is the purpose of `TaskInfo`?

Reply:
If only use the id or name directly in the code, it may be difficult to extend if we need to add more information about the task. So, I used a structure to describe the relevant information about the task.